plugins/nominal-connection-checker/src/index.js

```diff
@@ -42,21 +60,101 @@ export class NominalConnectionChecker extends Blockly.ConnectionChecker {
    */
   init(hierarchyDef) {
     this.typeHierarchy_ = new TypeHierarchy(hierarchyDef);
+    // TODO: Check if we need to release this.
+    this.workspace_.addChangeListener(this.onChangeListener_.bind(this));
   }
 
   /**
    * @override
    */
   doTypeChecks(a, b) {
     const {parent, child} = this.getParentAndChildConnections_(a, b);
-    // Checks should only contain a single type.
-    const parentType = parent.getCheck()[0];
-    const childType = child.getCheck()[0];
+    const parentType = this.getExplicitType_(parent);
+    const childType = this.getExplicitType_(child);
     const typeHierarchy = this.getTypeHierarchy_();
 
+    if (!parentType || !childType) {
+      // At least one is an unbound generic.
+      return true;
+    }
     return typeHierarchy.typeFulfillsType(childType, parentType);
   }
 
+  /**
+   * Listens to changes on the workspace, and handles things like binding and
+   * unbinding generic types to explicit types.
+   * @param {!Blockly.Event} e The current event.
+   * @private
+   */
+  onChangeListener_(e) {
+    if (e.type != Blockly.Events.BLOCK_MOVE) {
+      return;
+    }
+
+    const childBlock = this.workspace_.getBlockById(e.blockId);
+    const childCon = childBlock.outputConnection;
+    if (!childCon) {
+      // Ignore statement blocks for now;
+      return;
+    }
+
+    const genericMap = this.getGenericMap();
+    let parentBlock;
+    let parentCon;
+    let explicitFn;
+    let genericFn;
+    if (e.newParentId) {
+      parentBlock = this.workspace_.getBlockById(e.newParentId);
+      parentCon = parentBlock.getInput(e.newInputName).connection;
+      explicitFn = GenericMap.prototype.bindTypeToExplicit.bind(genericMap);
```

Comment:
Why are you using a bind instead of just calling the function, given that you also have the object around?

If the goal is only to make it possible to directly call the correct function, I think it would be clearer to set a boolean saying which case it is and then have a helper that checks that boolean and calls the appropriate function.

Reply:
Yeah this is a really tricky spot, and I still have mixed feelings about it.

The reason I originally aliased these functions instead of just calling them is because I wanted to try and keep the logic at the end of the function as simple as possible. All it cares about is the state of the connections (**E**, **GU**, or **GB**) so I wanted to move the other info about binding vs unbinding somewhere else.

Adding checks for binding vs unbinding at the end of the function would definitely work, but my current feeling is that it would distract from the logic that's really important. I know it's annoying to refer back to the actual function that's getting called, but the idea is that you shouldn't have to do that often because again, that info isn't really important. What's important is when it calls the generic function vs when it calls the explicit function.

That's just my current feeling on it though. As this function expands (especially as I add logic for flowing bindings through connections) something else may be better.